### PR TITLE
feat: Extract cccPolicy and do not create ccc triggers and accounts

### PR DESCRIPTION
**Pull request by doubleface**
Now all ccc specific code is gathered in cccPolicy.

Then ConnectionFlow can create an account, create a trigger, launch a trigger or not according to current konnector's policy

ccc specific code is still in `src/helpers/triggers.js` code since it will be used by flagship application to create accounts and triggers itself.

### CODE REVIEW

File: packages/cozy-harvest-lib/src/components/cards/LaunchTriggerAlert.jsx [hunk 1]
```diff
@@ -59,7 +59,8 @@ export const LaunchTriggerAlert = ({
   const styles = useStyles({ block })
 
   const lastSuccessDate = getLastSuccessDate(trigger)
-  const isKonnectorRunnable = isRunnable({ win: window, konnector })
+  const anchorRef = useRef()
+  const [showOptions, setShowOptions] = useState(false)
```
Comment (Crash--) [inline packages/cozy-harvest-lib/src/components/cards/LaunchTriggerAlert.jsx:63]: What is that?

File: packages/cozy-harvest-lib/src/konnector-policies.js [hunk 2]
```diff
@@ -1,6 +1,7 @@
 import logger from './logger'
 import { konnectorPolicy as biKonnectorPolicy } from './services/budget-insight'
 import { konnectorPolicy as biWebViewPolicy } from './services/biWebView'
+import { konnectorPolicy as cccPolicy } from './services/cccPolicy'
```
Comment (Crash--) [inline packages/cozy-harvest-lib/src/konnector-policies.js:4]: ccc means connecteur côté client, this is not the word we used. Name it clisk 🙏 . 

Also, other policies are namde `budget-insight` or `biWebview`. There is no `Policy` within the name.

File: packages/cozy-harvest-lib/src/konnector-policies.js [hunk 3]
```diff
@@ -26,8 +32,5 @@ logger.info('Available konnector policies', policies)
  * @param {KonnectorManifest} konnector
  * @returns {KonnectorPolicy}
  */
-export const findKonnectorPolicy = konnector => {
-  const policy = policies.find(policy => policy.match(konnector))
-  logger.info(`Using ${policy.name} konnector policy for ${konnector.slug}`)
-  return policy
-}
+export const findKonnectorPolicy = konnector =>
```
Comment (Crash--) [inline packages/cozy-harvest-lib/src/konnector-policies.js:35]: Why did you remove the log?

File: packages/cozy-harvest-lib/src/models/ConnectionFlow.js [hunk 4]
```diff
@@ -440,31 +440,51 @@ export class ConnectionFlow {
...
-      this.account = account
+        this.account = account
 
-      logger.info(`ConnectionFlow: Saved account ${account._id}`)
+        logger.info(`ConnectionFlow: Saved account ${account._id}`)
 
-      if (!konnectorPolicy.needsTriggerCreation) {
-        await this.ensureTriggerAndLaunch(client, {
+        logger.debug('ConnectionFlow: Ensuring trigger...')
+        const ensuredTrigger = await ensureTrigger(client, {
```
Comment (Crash--) [inline packages/cozy-harvest-lib/src/models/ConnectionFlow.js:460]: why this change in this commit?

Comment (Crash--) [reply packages/cozy-harvest-lib/src/models/ConnectionFlow.js:460]: Also, it should be called... `trigger` no?

File: packages/cozy-harvest-lib/src/models/ConnectionFlow.spec.js [hunk 5]
```diff
@@ -369,138 +369,67 @@ describe('ConnectionFlow', () => {
...
         trigger: fixtures.existingTrigger,
         userCredentials: fixtures.credentials
       })
 
-      expect(flow.ensureTriggerAndLaunch).toHaveBeenCalledWith(
+      expect(ensureTrigger).toHaveBeenCalledWith(
         client,
         expect.objectContaining({ account: fixtures.updatedAccount })
       )
     })
```
Comment (Crash--) [inline packages/cozy-harvest-lib/src/models/ConnectionFlow.spec.js:385]: Since now this `ensureTiggerAndLaunch` has been split into several call, you should also test that `ensureDefaultFolderPathInAccount` has been called for instance. 

Once it's done, I'm interested in what is the expected behavior if `ensureTrigger` resole, but `ensureDefaultFolderPathInAccount` throw.

File: packages/cozy-harvest-lib/src/models/ConnectionFlow.spec.js [hunk 6]
```diff
@@ -369,138 +369,67 @@ describe('ConnectionFlow', () => {
...
-      await flow.ensureTriggerAndLaunch(client, {
-        trigger: fixtures.createdTrigger,
-        konnector: fixtures.konnector
-      })
-      expect(launchTrigger).toHaveBeenCalledTimes(1)
-      expect(launchTrigger).toHaveBeenCalledWith(
-        client,
-        fixtures.createdTrigger
       )
     })
```
Comment (Crash--) [inline packages/cozy-harvest-lib/src/models/ConnectionFlow.spec.js:424]: You should also test the call to launch where this is not the : 
```
window.cozy = {
        ClientConnectorLauncher: 'react-native'
      }
```

File: packages/cozy-harvest-lib/src/models/ConnectionFlow.spec.js [hunk 7]
```diff
@@ -571,64 +501,91 @@ describe('ConnectionFlow', () => {
...
-      expect(saveAccount).not.toHaveBeenCalled()
+      expect(saveAccount).toHaveBeenCalledTimes(1)
       expect(client.query).toHaveBeenCalledWith(
         Q('io.cozy.files').getById(
           fixtures.createdTriggerWithFolder.attributes.message.folder_to_save
         )
       )
     })
+    delete window.cozy
+    delete window.ReactNativeWebView
```
Comment (Crash--) [inline packages/cozy-harvest-lib/src/models/ConnectionFlow.spec.js:526]: Where are these variables defined? Is it because of https://github.com/cozy/cozy-libs/pull/2000/commits/9f251f802b0dd046d1e99768b3b402350baed1b8#diff-57ee38fdf45b61888d99d54f38f4dfe8bc42c88c8e3af8e1026d120a1b9af17fR407-R411 ? 

If yes, this is a red flag. Each test must be autonomous and have no side effect

File: packages/cozy-harvest-lib/src/models/ConnectionFlow.spec.js [hunk 8]
```diff
@@ -571,64 +501,91 @@ describe('ConnectionFlow', () => {
...
+  })
+})
+
+describe('ensureTrigger', () => {
+  beforeAll(() => {
+    jest.spyOn(cronHelpers, 'fromFrequency').mockReturnValue('0 0 0 * * 0')
+  })
+
+  afterEach(() => {
+    jest.clearAllMocks()
```
Comment (Crash--) [inline packages/cozy-harvest-lib/src/models/ConnectionFlow.spec.js:588]: What is the goal of those lines?

File: packages/cozy-harvest-lib/src/services/cccPolicy.js [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+const isCCC = konnector => Boolean(konnector.clientSide)
+
+/**
+ * Get's the launcher in the current environment if any
+ *
+ * @param {Object} win The window object in the current environment
+ * @returns {Object|null}
+ */
+export const getLauncher = ({ win }) =>
+  win?.cozy?.ClientConnectorLauncher || null
```
Comment (Crash--) [inline packages/cozy-harvest-lib/src/services/cccPolicy.js:22]: Why do we need to pass the object in params? Can't we just read it from `window` directly?

File: packages/cozy-harvest-lib/src/konnector-policies.js [hunk 10]
```diff
@@ -9,10 +10,15 @@ const defaultKonnectorPolicy = {
   match: () => true,
   name: 'default',
   fetchExtraOAuthUrlParams: null,
-  refreshContracts: null
+  refreshContracts: null,
+  needsTriggerLaunch: true,
+  needsAccountAndTriggerCreation: true,
+  onLaunch: null,
+  isRunnable: () => true
```
Comment (Crash--) [inline packages/cozy-harvest-lib/src/konnector-policies.js:17]: `isRunnable` seems to take the `konnector` as arg. I prefer to see it there, same for `match` and else.

File: packages/cozy-harvest-lib/src/components/cards/LaunchTriggerAlert.jsx (hunk 1 above)
Comment (doubleface) [reply packages/cozy-harvest-lib/src/components/cards/LaunchTriggerAlert.jsx]: A mistake in rebase

File: packages/cozy-harvest-lib/src/konnector-policies.js (hunk 3 above)
Comment (doubleface) [reply packages/cozy-harvest-lib/src/konnector-policies.js:35]: This generates too many lines of logs in a simple connector execution. I now moved this removal in its own commit with explanations

File: packages/cozy-harvest-lib/src/konnector-policies.js (hunk 2 above)
Comment (doubleface) [reply packages/cozy-harvest-lib/src/konnector-policies.js:4]: Ok for clisk.

For the second part, I would prefer then to move the policies in their own directory named policies to make it clearer

File: packages/cozy-harvest-lib/src/konnector-policies.js (hunk 3 above)
Comment (Crash--) [reply packages/cozy-harvest-lib/src/konnector-policies.js:35]: Maybe it's a hint that there are too many call / re-renders ? Not sure that hiding it is the right solution. 

Is it intentional to have so many calls to this method?

File: packages/cozy-harvest-lib/src/models/ConnectionFlow.js (hunk 4 above)
Comment (doubleface) [reply packages/cozy-harvest-lib/src/models/ConnectionFlow.js:460]: I think it is clearer to separate account creation, trigger creation and trigger launch. And this makes it possible to policies to activated/deactivate any of them.

File: packages/cozy-harvest-lib/src/models/ConnectionFlow.spec.js (hunk 8 above)
Comment (doubleface) [reply packages/cozy-harvest-lib/src/models/ConnectionFlow.spec.js:588]: These lines are not actually needed anymore

File: packages/cozy-harvest-lib/src/models/ConnectionFlow.spec.js (hunk 7 above)
Comment (doubleface) [reply packages/cozy-harvest-lib/src/models/ConnectionFlow.spec.js:525]: Right, I put it back inside the same test now

File: packages/cozy-harvest-lib/src/services/cccPolicy.js (hunk 9 above)
Comment (doubleface) [reply packages/cozy-harvest-lib/src/services/cccPolicy.js]: Right, this is not actually needed. No reason to keep it

File: packages/cozy-harvest-lib/src/models/ConnectionFlow.spec.js [hunk 11]
```diff
@@ -571,64 +501,91 @@ describe('ConnectionFlow', () => {
...
       client.query.mockRejectedValue(new Error('404'))
-      await flow.ensureTriggerAndLaunch(client, {
+      await setupSubmit(flow, {
         account: fixtures.existingAccount,
         trigger: fixtures.existingTrigger,
         konnector: fixtures.konnectorWithFolder
       })
       expect(flow.account).toEqual(fixtures.updatedAccount)
-      expect(saveAccount).not.toHaveBeenCalled()
+      expect(saveAccount).toHaveBeenCalledTimes(1)
```
[the remark this answers is not in the document]
Comment (doubleface) [reply packages/cozy-harvest-lib/src/models/ConnectionFlow.spec.js:520]: The change is that we call flow.handleFormSubmit in this test instead of ensureTriggerAndLaunch. flow.handleFormSubmit calls createOrUpdateAccount, which is why a call to saveAccount is expected.

File: packages/cozy-harvest-lib/src/models/ConnectionFlow.spec.js (hunk 6 above)
Comment (doubleface) [reply packages/cozy-harvest-lib/src/models/ConnectionFlow.spec.js:426]: Done

File: packages/cozy-harvest-lib/src/models/ConnectionFlow.spec.js (hunk 5 above)
Comment (doubleface) [reply packages/cozy-harvest-lib/src/models/ConnectionFlow.spec.js:392]: Done.

Concerning `ensureDefaultFolderPathInAccount`, it depends on saveAccount, which retries on 409 errors and it does not throw (I fixed the saveAccount outside the try catch), errors are converted to warning messages. 

In this case (no default folder path defined), it can be defined once again on next account  update or the stack has a mecanism of defining a default folder with the name as the slug of the connector.

File: packages/cozy-harvest-lib/src/konnector-policies.js (hunk 3 above)
Comment (doubleface) [reply packages/cozy-harvest-lib/src/konnector-policies.js:36]: Since this method can be called with different konnectors  in the same session, it looks like it's mandatory to call it everytime we need a policy.

At first I thought we could define a an association object on load and use it for the next calls, but since the user can install, update konnectors in the interval, we cannot be sure that the same policy will be used for the same konnector everytime.

**Comment by doubleface** [comment]
@Crash-- A priori je suis à jour avec tes commentaires

File: packages/cozy-harvest-lib/src/components/cards/LaunchTriggerAlert.jsx [hunk 12]
```diff
@@ -57,7 +58,9 @@ export const LaunchTriggerAlert = ({
 
   const isInError = !!error
   const block = withDescription && (isInError || isInMaintenance)
-  const isKonnectorRunnable = isRunnable({ win: window, konnector })
+  const isKonnectorRunnable = findKonnectorPolicy(konnector).isRunnable({
```
[the remark this answers is not in the document]
Comment (Crash--) [reply packages/cozy-harvest-lib/src/components/cards/LaunchTriggerAlert.jsx:62]: @doubleface is it wanted to still have the args?

File: packages/cozy-harvest-lib/src/models/ConnectionFlow.js [hunk 13]
```diff
@@ -585,16 +582,15 @@ export class ConnectionFlow {
     try {
       const result = await client.query(Q('io.cozy.files').getById(folderId))
       folder = result.data
+      if (folder.path !== account.defaultFolderPath) {
+        account.defaultFolderPath = folder.path
+        const savedAccount = await saveAccount(client, konnector, account)
+        return savedAccount
+      }
```
Comment (Crash--) [inline packages/cozy-harvest-lib/src/models/ConnectionFlow.js:589]: can you explain why we do that inside the try / catch (aka why did you move the call)?

File: packages/cozy-harvest-lib/src/models/ConnectionFlow.spec.js [hunk 14]
```diff
@@ -369,138 +368,92 @@ describe('ConnectionFlow', () => {
...
-  describe('ensureTriggerAndLaunch', () => {
-    beforeAll(() => {
-      jest.spyOn(cronHelpers, 'fromFrequency').mockReturnValue('0 0 0 * * 0')
-      watchKonnectorJob.mockReturnValue({ on: () => ({}) })
-    })
-
-    afterEach(() => {
-      jest.clearAllMocks()
+      delete window.cozy
+      delete window.ReactNativeWebView
```
Comment (Crash--) [inline packages/cozy-harvest-lib/src/models/ConnectionFlow.spec.js:432]: can you test that `launch()` is not called?

File: packages/cozy-harvest-lib/src/models/ConnectionFlow.spec.js [hunk 15]
```diff
@@ -369,138 +368,92 @@ describe('ConnectionFlow', () => {
...
-        trigger: fixtures.existingTrigger,
-        konnector: fixtures.konnector
-      })
-      expect(launchTrigger).toHaveBeenCalledTimes(1)
-      expect(launchTrigger).toHaveBeenCalledWith(client, {
-        ...fixtures.existingTrigger
+      const promise = flow.handleFormSubmit({
+        konnector: fixtures.clientKonnector
       })
+      expect(async () => await promise).not.toThrow()
```
Comment (Crash--) [inline packages/cozy-harvest-lib/src/models/ConnectionFlow.spec.js:446]: but this time, should we test that `launch` is called?

File: packages/cozy-harvest-lib/src/models/ConnectionFlow.spec.js [hunk 16]
```diff
@@ -571,66 +525,83 @@ describe('ConnectionFlow', () => {
...
+        flow.konnector,
+        onAccountCreationResult
       )
     })
-    delete window.cozy
-    delete window.ReactNativeWebView
   })
+})
 
+describe('ensureTrigger', () => {
```
Comment (Crash--) [inline packages/cozy-harvest-lib/src/models/ConnectionFlow.spec.js:604]: is it wanted?

File: packages/cozy-harvest-lib/src/services/cccPolicy.js [hunk 17]
```diff
@@ -0,0 +1,66 @@
...
+ */
+
+import logger from '../logger'
+
+/**
+ * Check if the given konnector is a client side connector
+ * @param {import('cozy-client/types/types').KonnectorsDoctype} konnector - konnector object
+ * @returns {Boolean}
+ */
+const isCCC = konnector => Boolean(konnector.clientSide)
```
Comment (Crash--) [inline packages/cozy-harvest-lib/src/services/cccPolicy.js]: nit: should be inside the konnector model

File: packages/cozy-harvest-lib/src/components/cards/LaunchTriggerAlert.jsx (hunk 12 above)
[the remark this answers is not in the document]
Comment (doubleface) [reply packages/cozy-harvest-lib/src/components/cards/LaunchTriggerAlert.jsx:62]: no. This is now fixed

File: packages/cozy-harvest-lib/src/models/ConnectionFlow.js (hunk 13 above)
Comment (doubleface) [reply packages/cozy-harvest-lib/src/models/ConnectionFlow.js:613]: Just a better practice. To avoid to have uncaught errors and saveAccount can be a source of errors.

File: packages/cozy-harvest-lib/src/models/ConnectionFlow.spec.js (hunk 6 above)
Comment (Crash--) [reply packages/cozy-harvest-lib/src/models/ConnectionFlow.spec.js:528]: where?

File: packages/cozy-harvest-lib/src/models/ConnectionFlow.spec.js (hunk 16 above)
Comment (Crash--) [reply packages/cozy-harvest-lib/src/models/ConnectionFlow.spec.js:698]: cc @doubleface

File: packages/cozy-harvest-lib/src/models/ConnectionFlow.spec.js (hunk 15 above)
Comment (doubleface) [reply packages/cozy-harvest-lib/src/models/ConnectionFlow.spec.js:540]: If no launcher is available, the trigger is not launched

File: packages/cozy-harvest-lib/src/models/ConnectionFlow.spec.js (hunk 16 above)
Comment (doubleface) [reply packages/cozy-harvest-lib/src/models/ConnectionFlow.spec.js:698]: no, artifact of previous tests. Removed

File: packages/cozy-harvest-lib/src/models/ConnectionFlow.spec.js (hunk 15 above)
Comment (Crash--) [reply packages/cozy-harvest-lib/src/models/ConnectionFlow.spec.js:541]: I don't understand. If there is no launcher, then the policy will be the default one, right? And then, `needsTriggerLaunch` will be true, no?

Comment (doubleface) [reply packages/cozy-harvest-lib/src/models/ConnectionFlow.spec.js:541]: We are in case like this : in a web browser, in the home application, on a clisk.

- Since this is a web browser, there is no launcher available
- Since this is a clisk, the clisk policy is chosen

-> The trigger is not launched. Actually, the button to launch the konnector is not even enabled

I will even add a test to check that the trigger is not launched

Comment (Crash--) [reply packages/cozy-harvest-lib/src/models/ConnectionFlow.spec.js:542]: OK I got it, this is because of `konnector: fixtures.clientKonnector` 🤦

**Review by Crash-- — APPROVED** [review]
